src/generate/peripheral.rs

```diff
@@ -186,40 +145,237 @@ fn register_block(registers: &[Register], defs: &Defaults) -> Result<Tokens> {
+
+            if array_convertible {
+                register_expanded.push(RegisterBlockField {
+                    field: convert_svd_register(&register, name),
+                    description: info.description.clone(),
+                    offset: info.address_offset,
+                    size: register_size * array_info.dim,
+                });
+            } else {
+                let mut field_num = 0;
+                for field in expand_svd_register(register, name).iter() {
+                    register_expanded.push(RegisterBlockField {
+                        field: field.clone(),
+                        description: info.description.clone(),
+                        offset: info.address_offset + field_num * array_info.dim_increment,
+                        size: register_size,
+                    });
+                    field_num += 1;
+                }
+            }
+        }
+    }
+
+    Ok(register_expanded)
+}
+
+fn cluster_block(
+    c: &Cluster,
+    defaults: &Defaults,
+    p: &Peripheral,
+    all_peripherals: &[Peripheral],
+) -> Result<Tokens> {
+    let mut mod_items: Vec<Tokens> = vec![];
+
+    // name_sc needs to take into account array type.
+    let description = util::respace(&c.description);
+
+    // Generate the register block.
+    let mod_name = match *c {
+        Cluster::Single(ref info) => &info.name,
+        Cluster::Array(ref info, ref _ai) => &info.name,
+    }.replace("[%s]", "")
+        .replace("%s", "");
+    let name_sc = Ident::new(&*mod_name.to_sanitized_snake_case());
+    let reg_block = register_or_cluster_block(&c.children, defaults, Some(&mod_name))?;
+
+    // Generate definition for each of the registers.
+    let registers = util::only_registers(&c.children);
+    for reg in &registers {
+        mod_items.extend(register::render(
+            reg,
+            &registers,
+            p,
+            all_peripherals,
+            defaults,
+        )?);
+    }
+
+    // Generate the sub-cluster blocks.
+    let clusters = util::only_clusters(&c.children);
+    for c in &clusters {
+        mod_items.push(cluster_block(c, defaults, p, all_peripherals)?);
+    }
+
+    Ok(quote! {
+        #reg_block
+
+        /// Register block
+        #[doc = #description]
+        pub mod #name_sc {
+            #[allow(unused_imports)]
+            use vcell::VolatileCell;
+
+            #(#mod_items)*
+        }
+    })
+}
+
 /// Takes a svd::Register which may be a register array, and turn in into
 /// a list of syn::Field where the register arrays have been expanded.
-fn expand_svd_register(register: &Register) -> Vec<syn::Field> {
-    let name_to_ty = |name: &String| -> syn::Ty {
+fn expand_svd_register(register: &Register, name: Option<&str>) -> Vec<syn::Field> {
+    let name_to_ty = |name: &String, ns: Option<&str>| -> syn::Ty {
+        let ident = if let Some(ns) = ns {
+            Cow::Owned(
+                String::from("self::") + &ns.to_sanitized_snake_case() + "::"
```

Comment:
I don't understand why this change is needed, care to explain?

Reply:
This change allows for generation of names like this: https://gist.github.com/jamesmunns/03fd0d7d3595cd0816dbfa8daca0a553#file-ficr-diff-L58

Since clusters are nested namespaces, it is necessary for the outer (cluster) structures to refer to types inside of the nested namespaces.